File: resources/fmj_kt/src/fmj/goods/GoodsHiddenWeapon.kt
package fmj.goods

import fmj.characters.BuffMan
import fmj.lib.DatLib
import fmj.lib.ResBase
import fmj.lib.ResSrs

/**
 * 08暗器
 * @author Chen
 */
class GoodsHiddenWeapon : BaseGoods(), Throwable {

    /**
     *
     * @return 当该值为正时表示敌人损失多少生命，
     * 为负时表示从敌人身上吸取多少生命到投掷者身上
     */
    override var affectHp: Int = 0
        private set // 当该值为正时表示敌人损失多少生命，为负时表示从敌人身上吸取多少生命到投掷者身上
    /**
     *
     * @return 当该值为正时表示敌人损失多少真气，
     * 为负时表示从敌人身上吸取多少真气到投掷者身上
     */
    override var affectMp: Int = 0
        private set // 当该值为正时表示敌人损失多少真气，为负时表示从敌人身上吸取多少真气到投掷者身上
    override var ani: ResSrs = ResSrs()
        private set
    private var mBitMask: Int = 0 // 000 全体否 毒乱封眠

    override var buff = BuffMan()

    override fun setOtherData(buf: ByteArray, offset: Int) {
        affectHp = ResBase.get2BytesSInt(buf, offset + 0x16)
        affectMp = ResBase.get2BytesSInt(buf, offset + 0x18)
        val type = buf[offset + 0x1b].toInt() and 0xff
        val index = buf[offset + 0x1a].toInt() and 0xff
        ani = DatLib.getRes(DatLib.ResType.SRS, type, index) as ResSrs
        mBitMask = buf[offset + 0x1c].toInt() and 0xff
        buff = BuffMan.fromRoundAndMask(sumRound, mBitMask)
    }

    override fun effectAll(): Boolean {
        return mBitMask and 0x10 != 0
    }
}

File: packages/core/src/goods/goods-hidden-weapon.ts
import type { ResSrs } from '@/lib/res-srs';
import { STATUS_FLAG_ATTACK_ALL } from '@/characters/status';
import { BaseGoods, type BaseGoodsData } from './base-goods';

export interface GoodsHiddenWeaponData extends BaseGoodsData {
  readonly affectHp: number;
  readonly affectMp: number;
  readonly animation: ResSrs | null;
  readonly effectFlags: number;
}

export class GoodsHiddenWeapon extends BaseGoods {
  affectHp: number;
  affectMp: number;
  animation: ResSrs | null;
  effectFlags: number;

  constructor(data: GoodsHiddenWeaponData) {
    super(data);
    this.affectHp = data.affectHp;
    this.affectMp = data.affectMp;
    this.animation = data.animation;
    this.effectFlags = data.effectFlags;
  }

  override effectAll(): boolean {
    return (this.effectFlags & STATUS_FLAG_ATTACK_ALL) !== 0;
  }
}
